### ISSUE TITLE
RFE: the ability to checkpoint on a host and restore on another

### ISSUE THREAD

**Issue by muayyad-alsadi** [containers/libpod#1618]
Podman now [supports](https://podman.io/blogs/2018/10/10/checkpoint-restore.html) Checkpoint/Restore In Userspace (CRIU)
but it seems to be limited to a single host, that is to checkpoint and restore on same host.

how about checkpoint on a host, scp to another host and restore there?

**Comment by TomSweeneyRedHat** [comment]
@adrianreber Thoughts?

**Comment by adrianreber** [comment]
@muayyad-alsadi Like the linked article says, I am planing to implement migration support 'soon'. Not sure when 'soon' is, however. Hopefully I am able to open the necessary pull requests in the next few weeks.

**Comment by rhatdan** [comment]
@adrianreber Is looking into this.

**Comment by adrianreber** [comment]
I have a working proof of concept to migrate containers from one host to another working locally since today. I need to get the code into shape before submitting it, but it is working. It also needs runc changes which I will also look into.

**Comment by rhatdan** [comment]
This continues to move along.

**Comment by rhatdan** [comment]
@adrianreber Could you update this issue with  your progress.

**Comment by adrianreber** [comment]
The container migration feature required changes in many different projects. CRIU, runc, selinux-policy, container-selinux. Most problems have been solved. Right now I am waiting for the last changes in selinux-policy to be available on Fedora 28 and 29 so that the corresponding container-selinux policy changes to support the restore of multi-threaded processes with the correct SELinux labels.

I wrote a blog post about the things I had to change in CRIU to get everything working with SELinux: https://lisas.de/~adrian/posts/2019-Apr-10-criu-and-selinux.html

Right now all things seem to be solved and just need to be part of the corresponding distribution packages to be part of Podman's CI. Once it is available in the CI my pull request could turn green and then we can start the discussion about the actual code changes.

I would say we are pretty close to get this merged.
